OkHttp only connects to "h2" ALPN protocol

Values "h2-14", "h2-15", and "h2-16" result in [Protocol](https://github.com/square/okhttp/blob/293792c5ee66323c6070f9d88227d6579feaba47/okhttp/src/main/java/com/squareup/okhttp/Protocol.java#L87) throwing an error from [OkHttpTlsUpgrader](https://github.com/grpc/grpc-java/blob/master/okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java#L80).

We should support all those options, as we still are seeing them being used since the final HTTP/2 spec hasn't been published.

```
java.lang.RuntimeException: java.io.IOException: Unexpected protocol: h2-14
            at io.grpc.transport.okhttp.OkHttpClientTransport.start(OkHttpClientTransport.java:300)
            at io.grpc.ChannelImpl.obtainActiveTransport(ChannelImpl.java:188)
            at io.grpc.ChannelImpl.access$600(ChannelImpl.java:55)
            at io.grpc.ChannelImpl$CallImpl.start(ChannelImpl.java:248)
            at io.grpc.ForwardingCall.start(ForwardingCall.java:45)
...
     Caused by: java.io.IOException: Unexpected protocol: h2-14
            at com.squareup.okhttp.Protocol.get(Protocol.java:94)
            at com.squareup.okhttp.OkHttpTlsUpgrader.upgrade(OkHttpTlsUpgrader.java:80)
```

--- a/netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java
+++ b/netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java
@@ -68,10 +68,11 @@ import javax.net.ssl.SSLEngine;
  * endpoint.
  */
 public class Http2Negotiator {
-  // TODO(madongfly): Remove "h2-15" and "h2-16" at a right time.
+  // TODO(madongfly): Remove "h2-xx" at a right time.
   private static final List<String> SUPPORTED_PROTOCOLS = Collections.unmodifiableList(
       Arrays.asList(
           Http2OrHttpChooser.SelectedProtocol.HTTP_2.protocolName(),
+          "h2-14",
           "h2-15",
           "h2-16"));
 

--- a/okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java
+++ b/okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java
@@ -42,6 +42,7 @@ import java.net.ProxySelector;
 import java.net.Socket;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.List;
 
 import javax.net.ssl.SSLSocket;
 import javax.net.ssl.SSLSocketFactory;
@@ -51,6 +52,10 @@ import javax.net.ssl.SSLSocketFactory;
  * to do TLS upgrading.
  */
 public final class OkHttpTlsUpgrader {
+
+  // TODO(madongfly): We should only support "h2" at a right time.
+  private static final List<String> SUPPORTED_HTTP2_PROTOCOLS = Collections.unmodifiableList(
+      Arrays.asList("h2", "h2-14", "h2-15", "h2-16"));
 
   // A dummy address used to bypass null check.
   private static final InetSocketAddress DUMMY_INET_SOCKET_ADDRESS =
@@ -69,20 +74,26 @@ public final class OkHttpTlsUpgrader {
     spec.apply(sslSocket, getOkHttpRoute(host, port, spec));
 
     Platform platform = Platform.get();
-    try {
-      // Force handshake.
-      sslSocket.startHandshake();
+    // It's possible that the user provided SSLSocketFactory has already done the handshake
+    // when creates the SSLSocket.
+    String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
+    if (negotiatedProtocol == null) {
 
-      String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
-      if (negotiatedProtocol == null) {
-        throw new RuntimeException("protocol negotiation failed");
+      try {
+        // Force handshake.
+        sslSocket.startHandshake();
+
+        negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
+        if (negotiatedProtocol == null) {
+          throw new RuntimeException("protocol negotiation failed");
+        }
+      } finally {
+        platform.afterHandshake(sslSocket);
       }
-      Preconditions.checkState(Protocol.HTTP_2.equals(Protocol.get(negotiatedProtocol)),
-          "negotiated protocol is %s instead of %s.",
-          negotiatedProtocol, Protocol.HTTP_2.toString());
-    } finally {
-      platform.afterHandshake(sslSocket);
     }
+
+    Preconditions.checkState(SUPPORTED_HTTP2_PROTOCOLS.contains(negotiatedProtocol),
+        "negotiated protocol %s is unsupported", negotiatedProtocol);
 
     return sslSocket;
   }

